ethcontract/src/mock.rs

```diff
@@ -0,0 +1,65 @@
+//!
+
+use crate::json::Value;
+use crate::jsonrpc::Call;
+use crate::web3::{Error, RequestId};
+use web3::Transport;
+use crate::{Web3, Instance, Address};
+use crate::common::Abi;
+use crate::contract::MethodDefaults;
+use crate::dyns::{DynWeb3, DynTransport, DynInstance};
+use crate::tokens::Tokenize;
+use crate::common::hash::H32;
+
+pub struct Mock;
+impl Mock {
+    pub fn web3(&self) -> &DynWeb3 { todo!() }
+    pub fn transport(&self) -> &DynTransport { todo!() }
+    pub fn deploy(&self, abi: Abi) -> MockContract { todo!() }
+}
+
+pub struct MockContract;
+impl MockContract {
+    pub fn web3(&self) -> &DynWeb3 { todo!() }
+    pub fn transport(&self) -> &DynTransport { todo!() }
+    pub fn abi(&self) -> &Abi { todo!() }
+    pub fn address(&self) -> Address { todo!() }
+    pub fn defaults(&self) -> &MethodDefaults { todo!() }
+    pub fn defaults_mut(&mut self) -> &mut MethodDefaults { todo!() }
+    pub fn instance(&self) -> &DynInstance { todo!() }
+    pub fn instance_mut(&self) -> &mut DynInstance { todo!() }
+    pub fn into_instance(self) -> DynInstance { todo!() }
+
+    #[must_use]
+    pub fn expect_method<A, R>(&self, signature: H32) -> MockMethod<A, R> { todo!() }
+    #[must_use]
+    pub fn expect_view_method<A, R>(&self, signature: H32) -> MockMethodView<A, R> { todo!() }
+    #[must_use]
+    pub fn expect_fallback(&self) -> MockMethod<(), ()> { todo!() }
+
+    pub fn checkpoint(&mut self) { todo!() }
+}
+
+struct MockMethod<A: Tokenize, R: Tokenize>;
+impl<A, R> MockMethod<A, R> {
+    pub fn in_sequence(self, seq: mockall::Sequence) -> Self { todo!() }
+    pub fn times(self, times: impl Into<mockall::TimesRange>) -> Self { todo!() }
+    pub fn never(self) -> Self { todo!() }
+    pub fn once(self) -> Self { todo!() }
```

Comment:
I see that this looks a lot like some existing mock interfaces in rust, but would hope that we can also eventually include formal doc strings to each of this so that the auto-generated docs for this are approachable. Of course I assume this would have happened without my mentioning, but just wanted to say I would find this very beneficial.

Reply:
You're right, it follows `mockall` interface (which IMO could've been better, but oh well). And of course I'll add docs later, I just need to decide whether I want to mirror `mockall` or do a standalone thing =)

Reply:
I'd be keen to take this opportunity and change a few terms from mock all if your willing. Have you seen our solidity project MockContract? It's a bit old now and still in JS, but it might be worth seeing if the semantics there have any value here

https://github.com/gnosis/mock-contract

Reply:
Of course I also see the value in keeping things familiar and following in the footsteps of mockall